### PR TITLE
Add go pool for handling go routines

### PR DESCRIPTION
**Pull request by dhanalakshmi-everest**

### CODE REVIEW

File: pool.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	concurrency int
+	numCPUs     int
+	tasksChan   chan *Task
+	wg          sync.WaitGroup
+}
+
+var pool *Pool
+var once sync.Once
+
+func NewPool(tasks []*Task, concurrency int, numCPUs int) *Pool {
```
Comment (sammithk) [inline pool.go]: - Please make sure multiple clients will be able to send tasks after pool server starts.

File: pool.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	once.Do(func() {
+		pool = &Pool{
+			Tasks:       tasks,
+			concurrency: concurrency,
+			tasksChan:   make(chan *Task, concurrency),
+		}
+	})
+	pool.SetNumCPUs(numCPUs)
```
Comment (sammithk) [inline pool.go:29]: It would be better to use this option only while starting the pool server.

Comment (dhanalakshmi-everest) [reply pool.go:29]: Yeah but doing like that we won't be able to update the no of CPUs when creating new objects of it as it is a singleton object.

File: pool.go (hunk 1 above)
Comment (dhanalakshmi-everest) [reply pool.go]: Done

File: pool.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
+package Go_Pool
+
+import (
+	"errors"
+	"runtime"
+	"sync"
+)
+
+type Pool struct {
+	Tasks       []*Task
```
Comment (sammithk) [inline pool.go]: Since task are submitted via tasksChan, why is this field needed?

File: pool.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	runtime.GOMAXPROCS(numCPUs)
+	p.numCPUs = numCPUs
+
+	return nil
+}
+
+func (p *Pool) Run() {
+	for i := 0; i < p.concurrency; i++ {
```
Comment (sammithk) [inline pool.go]: why spawn all go routines at once, why not dynamically spawn them as and when required.

File: pool.go (hunk 3 above)
Comment (dhanalakshmi-everest) [reply pool.go]: I limited the tasksChan size to go routines count. To store extra tasks that need to executed whenever the tasksChan will get free will be stored in the Tasks slice.

File: pool.go (hunk 4 above)
Comment (dhanalakshmi-everest) [reply pool.go]: Done

File: pool.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+func NewPool(concurrency int, numCPUs int) *Pool {
+	once.Do(func() {
+		pool = &Pool{
+			concurrency:      concurrency,
+			tasksChan:        make(chan *Task, concurrency),
+			activeGoRoutines: make(chan bool, concurrency),
+		}
+	})
+	pool.SetNumCPUs(numCPUs)
+	go pool.schedule()
```
Comment (sammithk) [inline pool.go]: Is this required every time the function is called?

Comment (dhanalakshmi-everest) [reply pool.go]: Not required and updated it.

**Comment by sammithk** [comment]
Please satisfy all requirements for this diy.

**Comment by dhanalakshmi-everest** [comment]
> Please satisfy all requirements for this diy.

Done

**Comment by sammithk** [comment]
I think this requirement is missing

**Comment by sammithk** [comment]
this requirement is missing
Ability to drop requests after certain period if not served within a timeframe ( should be configurable by the client )

File: pool.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+func (p *Pool) work() {
+	for task := range p.tasksChan {
+		task.Run(&p.wg)
+	}
+}
+
+func (p *Pool) AddTask(task *Task) {
+	defer p.wg.Wait()
```
Comment (sammithk) [inline pool.go]: is this waitgroup required?

Comment (dhanalakshmi-everest) [reply pool.go]: Yeah, not needed and updated the changes.

**Review by sammithk — APPROVED** [review]

**Pull request merged**